tensorflow/contrib/layers/kernels/layer_norm_fused_op_gpu.cu.cc

```diff
@@ -0,0 +1,1455 @@
+/* Copyright 2015 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include <math.h>
+#include <stdio.h>
+#include <algorithm>
+
+#if GOOGLE_CUDA
+#define EIGEN_USE_GPU
+
+#include "./layer_norm_fused_op.h"
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+
+#if !defined(_MSC_VER)
+#define UNROLL _Pragma("unroll")
+#else
+#define UNROLL
+#endif
+
+#if !defined(__CUDA_ARCH__) || __CUDA_ARCH__ >= 600
+#else
+__device__ double atomicAdd(double* a, double b) { return b; }
+#endif
+
+#define MAX_GRID_SIZE 480
+inline int get_num_blocks(const int n_slices, const int slice_per_block) {
+  const int _num_blocks = n_slices / slice_per_block;
+  if (_num_blocks * slice_per_block == n_slices)
+    return _num_blocks;
+  else
+    return _num_blocks + 1;
+}
+
+inline int get_block_size(const int slice_size, int& block_size, int& mult) {
+  const int _warp_size = 32;
+  int _block_size = _warp_size;
+  int _mult = slice_size / _block_size;
+  mult = _mult * _block_size >= slice_size ? _mult : _mult + 1;
+  while (mult > 5) {
+    _block_size += _warp_size;
+    _mult = slice_size / _block_size;
+    mult = _mult * _block_size >= slice_size ? _mult : _mult + 1;
+  }
+  block_size = _block_size;
+  return 0;
+}
+
+template <typename T>
+__global__ void fillZeros(T* __restrict__ output, const int n_inputs) {
+  for (int thread_id = threadIdx.x + blockDim.x * blockIdx.x;
+       thread_id < n_inputs; thread_id += blockDim.x * gridDim.x)
+    output[thread_id] = static_cast<T>(0.0f);
+}
+
+template <typename T>
+__device__ __inline__ void warpSum(T& val1, T& val2) {
```

Comment:
In principle, this allows CUDA to issue two instructions at a time without being blocked by the previous `+=` operation. In reality, I am not really sure how much speed gain this would bring us. If you think this is not necessary, I am fine with two loops as well.